### ISSUE TITLE
New activations, e.g. and changed tReLU

### ISSUE THREAD

**Issue by PallHaraldsson** [FluxML/NNlib.jl#351]
I noticed your threshold version:

trelu(x, theta=1) = x > theta ? x : 0

and I certainly like the simplicity of it and ReLU.

I'm sure you're open to new functions (I could provide a PRs), but first a question on changing that one. I see it in a modified version used for winning results, it would be a breaking change (while users could hardcode in parameters to get old behaviour), do we need to keep the API in pre-1.0 version:

https://forums.fast.ai/t/comparison-of-new-activation-functions-on-imagenette-new-winner-trelu/44288/6
```
class TRelu(nn.Module):
    def __init__(self, threshold= - .25, mean_shift=-.03):
        super().__init__()
        self.threshold = threshold
        self.mean_shift = mean_shift
    
    def forward(self,x):
        x = F.relu(x)+self.threshold
        
        if self.mean_shift is not None:
            x.sub_(self.mean_shift)
            
        return x   
```

https://forums.fast.ai/t/comparison-of-activation-functions-2-winner-ftswish/44553

https://lessw.medium.com/comparison-of-activation-functions-for-deep-learning-initial-winner-ftswish-f13e2621847
Executive Summary: FTSwish with mean shifting (FTSwish+) was the top performing activation, and provided both highest net accuracy and one of the smoother training curves.

LiSHT: Non-Parametric Linearly Scaled Hyperbolic Tangent Activation Function for Neural Networks
https://arxiv.org/pdf/1901.05894.pdf
>The proposed LiSHT activation function
introduces more non-linearity in the network. It is completely
unbounded and solves the problems of diminishing gradient
problems. Other properties of LiSHT are symmetry, smooth-
ness and non-monotonicity, which play an important roles in
training.

Flatten-T Swish: a thresholded ReLU-Swish-like activation function for deep learning
https://arxiv.org/abs/1812.06247
>As compared with ReLU, FTS (T=-0.20) improves MNIST classification accuracy by 0.13%, 0.70%, 0.67%, 1.07% and 1.15% on wider 5 layers, slimmer 5 layers, 6 layers, 7 layers and 8 layers DFNNs respectively. Apart from this, the study also noticed that FTS converges twice as fast as ReLU.

**Comment by DhairyaLGandhi** [comment]
We could always return a closure to be non breaking

**Comment by ToucheSir** [comment]
Though this modified `trelu` does look interesting, I feel the threshold for inclusion into NNlib should be a tad more rigorous. The original author only tested on one dataset, and never got around to answering the very valid questions in https://forums.fast.ai/t/comparison-of-new-activation-functions-on-imagenette-new-winner-trelu/44288/11.

**Comment by PallHaraldsson** [comment]
A.
Neural Networks Fail to Learn Periodic Functions and How to Fix It
https://arxiv.org/pdf/2006.08195.pdf
>4 “Universal Extrapolation Theorem”
[..]
6.3 Financial Data Prediction

B.
Don't do anything rash with changing trelu. I just wanted to bring this up, to not forget, and also want to ask do you think it's actually much used?

I like this new LiSHT activation function I discovered, the numbers compared to ReLU, Swish and Tanh on many datasets, and the shape of the activation function itself:

https://github.com/swalpa/lisht

And here is a customized version of it:

https://github.com/lessw2020/LightRelu
>I implemented using Pytorch and wrapped it with a clamp and mean shift.(.46 and 7.5).
More testing in progress, but so far looks very promising!
Note - cut your learning rates in half vs ReLU, it learns very rapidly.

What I don't like about the activation function, is that it's more complex to compute, but that's ok since it learns rapidly (or does that only apply to the customized LightRelu?). I think I can make an approximate version of it (the original) that's very fast to evaluate, keeping it curved.

Many of the latest activation functions, Swish, Mish etc. are curved and non-monotonic. It makes sense that we need curved activation functions to approximate curved functions, but it also makes sense to have hard corners sometimes, as with ReLU. It seems to me, all the work on activation functions is a compromise, and we should rather have two types, curved and non-curved.

After writing all this I found what I added as section A. It seems only curved, i.e. periodic can work. I still think maybe having two types, or maybe then three, could help. I was thinking, possibly a bit choosing the activation function to use per neuron... Or at least is there something fundamental about using the same one across layers?
